### DISCUSSION TITLE
Where is the Javascript console?

### DISCUSSION THREAD

**Discussion by helander** [pocketbase/pocketbase#3024] [Q&A] [answered]
I am running verson 0.17.0 and tested the JS example with the hello route. I can make a curl call to the route and it answers with Hello and name. The code calls the console.log() but where can this output be found?

**Comment by sergio9929** [comment]
It logs into the console from which you executed Pocketbase.

**Reply by helander** [reply]
No, it does not for me. Do I need to do something to enable it?

**Reply by ganigeorgiev** [reply]
@helander Could you provide a code sample? Are you sure that your route is actually being executed and not triggering some other action?

**Reply by helander** [reply]
Sorry, I was not checking the code enough, the console.log call was not in a route but in an onModel… handler.
Now it works, thanks.

**Reply by helander** [reply]
I also found out that the console is written to the stderr of the pocketbase process.

Is any of this in the documentation?

**Reply by helander** [reply]
In the documentation there is a section about extending the console commands. The documentation describes how you could add your own custom commands, but how do you actually trigger the execution of these commands? 
Could you somehow connect to the pocketbase process and enter commands?

**Reply by ganigeorgiev** [reply]
@helander

> I also found out that the console is written to the stderr of the pocketbase process. ...

It is to the default logger, equivalent to [`log.Print`](https://pkg.go.dev/log#Print). Where else you'll like it to be written? If you are asking for custom logging, this will be handled separately as part of planned "Logs generalization" tasks in the future.

> The documentation describes how you could add your own custom commands, but how do you actually trigger the execution of these commands?

I'm not sure I follow. The [custom console commands](https://pocketbase.io/docs/js-console-commands/) are expected to be executed from the developers via cli (eg. `./pocketbase hello`, _I'll update later the example to be more clear_).

> Could you somehow connect to the pocketbase process and enter commands?

What you are looking for probably is https://pocketbase.io/docs/js-routing/#registering-new-routes.

**Reply by helander** [reply]
Regarding the log output, I have no further needs. I observed that the JS console output showed up on stderr (and not on stdout) beacuse I split this into different log files in my systemd unit.

So the extra console commands are to be executed by another process than the pocketbase ”serve” process ? My thinking was that you only had a single process ( The SERVER ) and you should connect to that process somehow. I now assume that it was not correct, and that custom commands are executed in additional processes created as part of the command invocation, correct?

Can these processes ”operate” on the same pb_data directory ”simultaneously”, i.e. Is the data directory protected from parallel execution (can I run a console command while server clients update the data via the serve process?

**Reply by ganigeorgiev** [reply, marked as answer]
> custom commands are executed in additional processes created as part of the command invocation, correct?

Yes. This is true also if you register a custom console command via Go. Examples for custom console commands are the `admin`, `update` (_works only with the prebuilt executable_) and `migrate` builtins.

> Can these processes ”operate” on the same pb_data directory ”simultaneously”?

Yes, but each PocketBase process have their own app context, meaning that the event hook from one process are not shared with the other. If you need something to be executed as part of the current running PocketBase instance, a custom rpc-style route would be more appropriate in this case.

**Reply by ganigeorgiev** [reply]
>  I observed that the JS console output showed up on stderr

Yes, that's correct. `console.*` uses the default Go logger for everything which output to stderr (see https://github.com/golang/go/blob/master/src/log/log.go#L87) and I think this is the correct behavior because stderr is used usually for diagnostic purposes.

I guess this may differ from node where I think `console.log` goes to stdout and `console.error` to stderr (_I haven't verified it_) but ~this will complicate the implementation of the [goja console module](https://github.com/dop251/goja_nodejs/blob/master/console/module.go) (especially if a custom printer is used) and I'm not sure if it is worth opening a PR for it~ actually it may not be a bad idea to overwrite the default goja printer; I'll consider experimenting to see how feasible would be to make it more consistent with node.

My recommendation would be to use `console.*` only as part of the development for quick "checks". A better logging (with option to write and explore the result in the Admin UI) will be available in the near future.

**Reply by helander** [reply]
Another question related to processes is about Job Scheduling. I assume this only works well within the ”serve” process and not in short-living custom command processes. I assume that you could have a custom command that never return (no return from the command run() function and in that process define scheduled jobs?

**Reply by ganigeorgiev** [reply]
>  I assume this only works well within the ”serve” process and not in short-living custom command processes. 

Currently the JSVM cron ticker is initialized with the [`onAfterBootstrap` event](https://pocketbase.io/docs/js-event-hooks/#onafterbootstrap) (or if `cronAdd` is called manually after the app bootstrap) but this may change in the future since as you've noted, the console commands are not a long-running process and I'm not sure whether if the job scheduling is useful in this case as the command may not run long enough to trigger the cron handler(s). And depending on the cron expression actually it may trigger the cron handler multiple times if the serve process is also running which may not be expected by the users. I'll think a little more on it and eventually may disable the JSVM cron ticker for non-serve processes (aka. start the cron ticker if only `OnBeforeServe` is triggered).

**Reply by helander** [reply]
I can see the value of not having the scheduled jobs in the serve process for stability purposes (in my mind the serve process may have higher stability requirements than scheduling jobs), so I would vote for having the option to use scheduled jobs outside of the serve process. I am not opposing to have jobs in the serve process.

**Reply by helander** [reply]
In order to have scheduled jobs in a console command process you need to have that process somehow suspend the command execution ”forever”. Is that possible to do?

**Reply by ganigeorgiev** [reply]
No, currently we don't expose such binding and I'm not sure if it will be a good idea to implement it since it could cause more harm than good.

Earlier today I've disabled the cron ticker for the case where `OnBofeoreServe` is not triggered (https://github.com/pocketbase/pocketbase/commit/9254ce46eb19f86c84cac47d6b87e8f2b4f6edbb) and will consider pushing it as v0.17.1 sometime later this week (_I'm waiting for the resolution of https://github.com/dop251/goja_nodejs/pull/60 to decide whether to implement a custom printer or to wait the upstream_).

The documentations will be also updated to reflect the change.

**Reply by helander** [reply]
I came up with another way to run scheduled jobs outside of the http/https server process. If you start a serve process and provide empty values for —http and —https, it looks like you get a process without the http/ https ”listeners”. Maybe this is a way forward an then you could disallow jobs in console command processes.

**Reply by ganigeorgiev** [reply]
But why would you want to do it anyway?

Keep in mind that the scheduled jobs run in their own goroutine in a non-blocking manner so it is completely fine to run alongside the `serve` process.

**Reply by helander** [reply]
Goroutine isolation is quite different than process isolation from a stability point of view, but it is not super important for me in my current project. But in other projects the view may be different …

**Reply by ganigeorgiev** [reply]
The HTTP serving process is expected to be long-running and at least for me should have higher priority from stability point of view because that's actually your backend. I can't think of a scenario where you'll want to execute PocketBase scheduled job but the http server is not running. For such use case the OS-level cron utility should be preferred.

The JSVM comes with some limitations (see also https://github.com/pocketbase/pocketbase/discussions/2825) and there will be always someone unhappy because of the go-js bridge. If your use case require more control over the execution flow you can try exploring the [Go APIs](https://pocketbase.io/docs/go-jobs-scheduling/). When using Go you don't even need to use the [PocketBase cron package](https://pkg.go.dev/github.com/pocketbase/pocketbase/tools/cron) and can initialize any other 3rd party library.

**Reply by helander** [reply]
My scheduled jobs are operating towards the database, and not having to require a network interface towards the database and at the same time have great isolation - because the database server are more important than the jobs. 
I now understand that if I want isolation then my jobs would have to interface to the database via a network based protocol, I will live with that.

**Reply by ganigeorgiev** [reply]
> I now understand that if I want isolation then my jobs would have to interface to the database via a network based protocol...

I'm not sure if this is the correct conclusion that should be taken, but at the same time I'm also not sure that I understand what do you refer to by _isolation_ in this context.

The database (SQLite) is always part of the application process. There is no separate db server running. If you want to interact with the database without requiring the HTTP server to be running then you can create console commands. The scheduled jobs require a long running process. We just "piggyback" to the already long-running process that your application most likely has (aka. the http server). As a bonus, since the scheduled jobs runs in the same process as the web server they share the same app event hooks so basically this allow communicating with each other (aka. the realtime events will work if for example you modify a record in a scheduled job, as this is the behavior most users will probably expect).

We can add JSVM binding to allow creating long running blocking tasks by registering a console command, but again I'm reluctant to introduce it at the moment as it may cause more harm than good.

**Reply by helander** [reply]
In do not want to have this discussion become a long running thing 😉
As I have said my main issue is isolation and what that means for me is partly this (there are other parts that stems from how to split your system up in different executing units; processes, threads, goroutines, …)
One important factor in order to reach a certain level of availability of the system, is how errors may propagate between executing units. An error in a thread or goroutine can impact all other units hosted in the same process as the unit where the error occured. In worst case the whole process will have to be restarted. However errors can not impact units in other processes in this drastic way, due to the hardware memory protection the OS sets up between processes. If you isolate units from each other, you lower the impacts of errors. This is what I mean by ”isolation”.
